multi_partner_learning.py

```diff
@@ -156,52 +236,75 @@ def compute_test_score(self):
 
                 self.minibatch_index = minibatch_index
 
-                if self.learning_approach == 'fedavg':
+                if self.learning_approach == "fedavg":
                     self.compute_collaborative_round_fedavg()
 
-                elif self.learning_approach in ['seq-pure', 'seq-with-final-agg']:
-                    self.compute_collaborative_round_sequential(sequentially_trained_model)
+                elif self.learning_approach in ["seq-pure", "seq-with-final-agg"]:
+                    self.compute_collaborative_round_sequential(
+                        sequentially_trained_model
+                    )
 
-                elif self.learning_approach == 'seqavg':
+                elif self.learning_approach == "seqavg":
                     self.compute_collaborative_round_seqavg()
+                
+                elif self.learning_approach == "stacking":
+                    self.compute_collaborative_round_stacking(stacking_model)
```

Comment:
@Thomas-Galtier there is something I don't get 😃 Why do you train the meta-model with this custom `epochs x mini-batches` approach, why not just 1 single `.fit()`?

Reply:
Well training the meta-model is like training the main model in other approaches ( sequential or fedavg ). The `epoch x minibatch` approach should  improve the quality of the fit of the model, especially in stratified cases.

Reply:
Well, in this case it seems to me that computation-wise it doesn't present an advantage versus the other approaches.

My understanding before was that:

- we do an independent training on each partner's data; this costs 1 'full-scale' computation
- then for any coalition of partners that we want to assess, we can assemble their models in a meta-model; this costs not much
- this way, a Shapley-based approach would be less costly, as it wouldn't require 'full-scale' training for each coalition to assess

But eventually if for a given coalition of partners we have to do a 'full scale' training on their data, this rationale doesn't hold. Sorry if it was clear for some time, I just figured this out now.

Some questions @Thomas-Galtier :

- So in a sense, we could see the approach you implemented as starting with a pre-trained model? The assembly of the individual models create a kind of pre-trained model, which we train on all data?
- Or, in a Shapley-like approach where we need to assess different coalition of partners, could we do the full-scale training only for the larger coalition, and derive the other ones from what we learned here (without the need to re-train the meta-model)?

Reply:
* My hope was that the full-scale training on the meta-model would be quicker than  a full-scale training with the main model, because the meta model is lighter. But It does not seems to be the case. I  guess   that tensorflow still computes the gradients  of all the architecture of the stacked model ( for each  partner's  model  and for the meta-model), and then only apply the gradient  of the meta model part (as the rest of the architecture is set to not trainable variables). Thus it computes a lot of gradients that it does not uses in the end, that's why it is so slow to train the meta model...   I may have an idea to avoid theses extra computations : Instead of training the stacked model we can train only the meta model by setting its input as the ouputs of each partner model.

* No the approach cannot be seen as starting with a pretrained model, as the architecture of the stacked model (N models stacked in parallel + a meta model to aggregate their outputs) is different from the architecture of a single partner model .

* I don't know how  we can use the stacked model on the larger coalition to infer what would be a stacked model on a smaller coalition as the two stacked model on each coalition do not have the same architecture.